crates/matrix-sdk-crypto/src/backups/mod.rs

```diff
@@ -156,26 +268,76 @@ impl BackupMachine {
-                            if device.verified()
-                                && device.is_signed_by_device(serialized_auth_data.clone()).is_ok()
-                            {
-                                return Ok(true);
-                            }
-                        } else {
-                            trace!(
-                                device_id = device_key_id.device_id().as_str(),
-                                "Device not found, can't check signature"
-                            );
+                        let state =
+                            self.backup_signed_by_device(device_key_id, device, auth_data.clone());
+
+                        result.insert(device_key_id.device_id().to_owned(), state);
+
+                        // Abort the loop if we found a trusted and valid
+                        // signature, unless we should check all of them.
+                        if state.trusted() && !check_all {
+                            break;
                         }
                     }
                 }
             }
+        }
+
+        Ok(result)
+    }
+
+    async fn verify_auth_data_v1(
+        &self,
+        auth_data: MegolmV1AuthData,
+        check_all: bool,
+    ) -> Result<SignatureCheckResult, CryptoStoreError> {
+        trace!(?auth_data, "Verifying backup auth data");
+        let serialized_auth_data = serde_json::to_value(auth_data.clone())?;
+
+        // Check if there's a signature from our own device.
+        let device_signature = self.check_own_device_signature(serialized_auth_data.clone()).await;
+        // Check if there's a signature from our own user identity.
+        let user_identity_signature =
+            self.check_own_identity_signature(serialized_auth_data.clone()).await?;
+
+        // Collect all the other signatures if there isn't already a valid one,
+        // or if we're told to collect all of them anyways.
+        let other_signatures =
+            if !device_signature.trusted() || !user_identity_signature.trusted() || check_all {
+                self.backup_signed_by_any_trusted_device(
+                    &auth_data.signatures,
+                    serialized_auth_data,
+                    check_all,
+                )
+                .await?
+            } else {
+                Default::default()
+            };
+
+        Ok(SignatureCheckResult { device_signature, user_identity_signature, other_signatures })
+    }
+
+    /// Verify some backup auth data that we downloaded from the server.
+    ///
+    /// The auth data should be fetched from the server using the
+    /// [`/room_keys/version`] endpoint.
+    ///
+    /// [`BackupAlgorithm`]: ruma::api::client::backup::BackupAlgorithm
+    /// [`/room_keys/version`]: https://spec.matrix.org/unstable/client-server-api/#get_matrixclientv3room_keysversion
+    pub async fn verify_backup(
+        &self,
+        backup_version: Value,
+        check_all: bool,
```

Comment:
`check_all` isn't documented here. And I don't really like its existence, if I'm being honest. :'(

For other readers, context on why `check_all` exists is [here](https://github.com/matrix-org/matrix-rust-sdk/issues/709).

Reply:
Done in https://github.com/matrix-org/matrix-rust-sdk/pull/725/commits/b996c06e4e8d67861f3b23fec9690aed814cb133. As well as some more name changes that were suggested in chat.